File: StriverA2Z/C++/3_ArraysQuestion/3.2_Medium/reverse_matrix_90_degree.cpp
#include<bits/stdc++.h>

using namespace std;

void printMatrix(vector<vector<int>> vec) {
    for (int i = 0; i < vec.size(); i++) {
        for (int j = 0; j < vec[0].size(); j++) {
            cout << vec[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl;
}

void reverseArray(vector<int> &arr, int start, unsigned long long end) {
    while (start <= end) {
        int temp = arr[start];
        arr[start] = arr[end];
        arr[end] = temp;
        start++;
        end--;
    }
}

void reverse_90_degree(vector<vector<int>> &matrix) {

    printMatrix(matrix);

    for (int i = 0; i < matrix.size() - 1; i++) {
        for (int j = i + 1; j < matrix[0].size(); j++) {
            int temp = matrix[i][j];
            matrix[i][j] = matrix[j][i];
            matrix[j][i] = temp;
        }
    }

    for (vector<int> &i: matrix) {
        reverseArray(i, 0, i.size() - 1);
    }

    printMatrix(matrix);
}

int main() {
    vector<vector<int>> vec = {{1, 2, 3},
                               {4, 5, 6},
                               {7, 8, 9}};
    reverse_90_degree(vec);
    return 0;
}

File: StriverA2Z/Java/Arrays/Medium/ReverseMatrix90Degree.java
public class ReverseMatrix90Degree {
    public static void main(String[] args) {
        int[][] arr = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
        reverse90Degree(arr);
    }

    static void reverse90Degree(int[][] arr) {
        printMatrix(arr);

        for (int i = 0; i < arr.length; i++) {
            for (int j = i + 1; j < arr[0].length; j++) {
                int temp = arr[i][j];
                arr[i][j] = arr[j][i];
                arr[j][i] = temp;
            }
        }

        for (int i = 0; i < arr.length; i++) {
            reverseArray(arr[i], 0, arr[i].length - 1);
        }

        printMatrix(arr);
    }

    static void reverseArray(int[] arr, int start, int end) {
        while (start <= end) {
            int temp = arr[start];
            arr[start] = arr[end];
            arr[end] = temp;
            start++;
            end--;
        }
    }

    static void printMatrix(int[][] arr) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr[0].length; j++) {
                System.out.print(arr[i][j] + " ");
            }
            System.out.println();
        }
        System.out.println();
    }
}
